libs/core/type_support/include/hpx/type_support/relocate_at.hpp

```diff
@@ -0,0 +1,101 @@
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+#include <hpx/type_support/construct_at.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+namespace hpx {
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+    using std::relocate_at;    // what would be the include for this?
+    using std::relocate;
+#else
+
+    namespace detail {
+        template <class T>
+        struct destroy_guard
+        {
+            T& t;
+            explicit destroy_guard(T& t)
+              : t(t)
+            {
+            }
+            ~destroy_guard()
+            {
+                t.~T();
+            }
+        };
+
+        template <class St, class Dt,
+            std::enable_if_t<
+                std::is_same_v<std::decay_t<St>, std::decay_t<Dt>> &&
+                hpx::is_trivially_relocatable_v<std::decay_t<St>> &&
+                !std::is_volatile_v<St> && !std::is_volatile_v<Dt>,
+                int> = 0>
+        Dt* relocate_at_helper(St* src, Dt* dst) noexcept
+        {
+            std::memmove(dst, src, sizeof(St));
+            return std::launder(dst);
+        }
+
+        // this is move and destroy
+        template <typename St, typename Dt,
+            std::enable_if_t<
+                !std::is_same_v<std::decay_t<St>, std::decay_t<Dt>> ||
+                !hpx::is_trivially_relocatable_v<std::decay_t<St>> ||
+                std::is_volatile_v<St> || std::is_volatile_v<Dt>,
+                int> = 0>
+        Dt* relocate_at_helper(St* src, Dt* dst) noexcept(
+            // has non-throwing move constructor
+            std::is_nothrow_constructible_v<Dt, St&&>)
+        {
+            destroy_guard<St> g(*src);
+            return hpx::construct_at(dst, HPX_MOVE(*src));
+        }
+    }    // namespace detail
+
+    template <typename T>
+    T* relocate_at(T* src, T* dst) noexcept(
+        std::is_nothrow_move_constructible_v<T>)
+    {
+        static_assert(
+            std::is_move_constructible_v<T> && std::is_destructible_v<T>,
+            "construct_at(dst, std::move(*source)) must be well-formed");
+
+        return hpx::detail::relocate_at_helper(src, dst);
+    }
+
+/*  ABI HACKY PART
+    template <class T,
+        typename std::enable_if_t<!hpx::is_trivially_relocatable_v<T> ||
+                std::is_trivially_destructible_v<T>,
+            int> = 0>
+    T relocate(T* source)
+    {
+        auto g = hpx::detail::destroy_guard(*source);
+        return HPX_MOVE(*source);
+    }
+
+    template <class T,
+        typename std::enable_if_t<hpx::is_trivially_relocatable_v<T> &&
+                !std::is_trivially_destructible_v<T>,
+            int> = 0>
+    T relocate(T* source)
+    {
+        auto magic = (T(*)(void*, size_t)) memcpy;
+        return magic(source, sizeof(T));
+    }
+    // more info this https://quuxplusone.github.io/blog/2022/05/18/std-relocate/
+*/
```

Comment:
> I was wondering, do you think there's a standard-friendly way to achieve the same thing?

The spec's reference implementation doesn't get you the memmove codegen, but it's perfectly conforming:
```
std::remove_cv_t<T> t = std::move(source);
std::destroy_at(source);
return t;
```
There is no way to get the memmove codegen without bending some rule or other: either use my hack, or, here's an alternative direction that's been discussed:
```
    __attribute__((do_not_construct)) T t;
    std::relocate_at(source, std::address_of(t));
    return t;  // NRVO
```
But no compiler vendor has ever implemented such an attribute. (I don't even think anyone's implemented [P1247 `__attribute__((no_destroy))`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1247r0.html).) So unless you own your own compiler, that direction is currently a dead end. :)

Reply:
Sad to hear, does P1144 require compilers to implement a way to get the memmove codegen, or is not required as per the as-if rule?

Reply:
> does P1144 require compilers to implement a way to get the memmove codegen

Nope. This is a deliberate design decision, both
- pragmatically — I don't want vendors to _have_ to jump through every single hoop on the first try; let them deal with performance as a quality-of-implementation issue at their leisure
- philosophically — the implementation _may_ skip observable side effects associated with trivial relocation; but if you're _relying_ on the implementation to skip a side effect of relocation because you know that that side effect would change the semantics of your program, well then, it sounds like your type isn't trivially relocatable after all!